File: macOS/Packages/ClaudeUsageCore/Sources/ClaudeUsageCore/AIToolParsers.swift
import Foundation

public enum GeminiSessionParser {
    public static func parse(_ data: Data, fallbackSessionId: String = "unknown") -> [AIToolUsageEvent] {
        if let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            return parseLegacy(root, fallbackSessionId: fallbackSessionId)
        }
        guard let text = String(data: data, encoding: .utf8) else { return [] }
        var sessionId = fallbackSessionId
        var projectPath = "unknown"
        var events: [AIToolUsageEvent] = []
        for line in text.split(whereSeparator: \.isNewline) {
            guard let lineData = String(line).data(using: .utf8),
                  let object = try? JSONSerialization.jsonObject(with: lineData) as? [String: Any] else { continue }
            sessionId = string(object["sessionId"]) ?? sessionId
            projectPath = firstDirectory(object) ?? projectPath
            if let set = object["$set"] as? [String: Any] {
                sessionId = string(set["sessionId"]) ?? sessionId
                projectPath = firstDirectory(set) ?? projectPath
                if let messages = set["messages"] as? [[String: Any]] {
                    events.append(contentsOf: messages.compactMap { event($0, sessionId: sessionId, projectPath: projectPath) })
                }
            }
            if let parsed = event(object, sessionId: sessionId, projectPath: projectPath) { events.append(parsed) }
        }
        return deduplicate(events)
    }

    private static func parseLegacy(_ root: [String: Any], fallbackSessionId: String) -> [AIToolUsageEvent] {
        let sessionId = string(root["sessionId"]) ?? fallbackSessionId
        let projectPath = firstDirectory(root) ?? "unknown"
        let messages = root["messages"] as? [[String: Any]] ?? []
        return deduplicate(messages.compactMap { event($0, sessionId: sessionId, projectPath: projectPath) })
    }

    private static func event(_ value: [String: Any], sessionId: String, projectPath: String) -> AIToolUsageEvent? {
        guard string(value["type"]) == "gemini",
              let tokens = value["tokens"] as? [String: Any],
              let timestampText = string(value["timestamp"]),
              let timestamp = parseDate(timestampText) else { return nil }
        let input = integer(tokens["input"])
        let output = integer(tokens["output"])
        let cached = integer(tokens["cached"])
        let reasoning = integer(tokens["thoughts"])
        let total = integer(tokens["total"])
        guard total > 0 || input > 0 || output > 0 else { return nil }
        let messageId = string(value["id"]) ?? "\(timestamp.timeIntervalSince1970)"
        return AIToolUsageEvent(
            timestamp: timestamp,
            tool: .geminiCLI,
            provider: "google",
            model: string(value["model"]) ?? "unknown",
            inputTokens: input,
            outputTokens: output,
            cachedTokens: cached,
            reasoningTokens: reasoning,
            totalTokens: total > 0 ? total : input + output + reasoning,
            eventId: "gemini:\(sessionId):\(messageId)",
            sessionId: sessionId,
            projectPath: projectPath
        )
    }

    private static func firstDirectory(_ value: [String: Any]) -> String? {
        (value["directories"] as? [String])?.first
    }

    private static func deduplicate(_ events: [AIToolUsageEvent]) -> [AIToolUsageEvent] {
        Array(Dictionary(events.map { ($0.eventId, $0) }, uniquingKeysWith: { _, latest in latest }).values)
    }
}

public enum OpenCodeMessageParser {
    /// Parses the JSON stored in OpenCode's `message.data` SQLite column. User messages
    /// and response text are ignored; only the assistant usage envelope is retained.
    public static func parseMessageData(_ data: Data) -> AIToolUsageEvent? {
        guard let value = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              string(value["role"]) == "assistant",
              let tokens = value["tokens"] as? [String: Any],
              let time = value["time"] as? [String: Any] else { return nil }
        let created = double(time["created"])
        guard created > 0 else { return nil }
        let input = integer(tokens["input"])
        let output = integer(tokens["output"])
        let reasoning = integer(tokens["reasoning"])
        let cache = tokens["cache"] as? [String: Any]
        let cached = integer(cache?["read"])
        let explicitTotal = integer(tokens["total"])
        let total = explicitTotal > 0 ? explicitTotal : input + output + reasoning
        guard total > 0 else { return nil }
        let messageId = string(value["id"]) ?? "\(created)"
        let sessionId = string(value["sessionID"]) ?? "unknown"
        let path = value["path"] as? [String: Any]
        return AIToolUsageEvent(
            timestamp: Date(timeIntervalSince1970: created / 1000),
            tool: .openCode,
            provider: string(value["providerID"]) ?? "unknown",
            model: string(value["modelID"]) ?? "unknown",
            inputTokens: input,
            outputTokens: output,
            cachedTokens: cached,
            reasoningTokens: reasoning,
            totalTokens: total,
            eventId: "opencode:\(messageId)",
            sessionId: sessionId,
            projectPath: string(path?["cwd"]) ?? "unknown"
        )
    }
}

private func string(_ value: Any?) -> String? { value as? String }
private func integer(_ value: Any?) -> Int {
    if let number = value as? NSNumber { return number.intValue }
    if let text = value as? String { return Int(text) ?? 0 }
    return 0
}
private func double(_ value: Any?) -> Double {
    if let number = value as? NSNumber { return number.doubleValue }
    if let text = value as? String { return Double(text) ?? 0 }
    return 0
}
private func parseDate(_ value: String) -> Date? {
    let fractional = ISO8601DateFormatter()
    fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    return fractional.date(from: value) ?? ISO8601DateFormatter().date(from: value)
}

File: TokenMihariban/Parsing/AIToolParsers.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text.Json;
using TokenMihariban.Models;

namespace TokenMihariban.Parsing;

public static class GeminiSessionParser
{
    public static IReadOnlyList<AIToolUsageEvent> Parse(byte[] data, string fallbackSessionId = "unknown")
    {
        try
        {
            using var doc = JsonDocument.Parse(data);
            if (doc.RootElement.ValueKind == JsonValueKind.Object)
                return ParseLegacy(doc.RootElement, fallbackSessionId);
        }
        catch { }

        var sessionId = fallbackSessionId;
        var projectPath = "unknown";
        var result = new List<AIToolUsageEvent>();
        foreach (var line in System.Text.Encoding.UTF8.GetString(data).Split('\n', StringSplitOptions.RemoveEmptyEntries))
        {
            try
            {
                using var doc = JsonDocument.Parse(line);
                var root = doc.RootElement;
                sessionId = String(root, "sessionId") ?? sessionId;
                projectPath = FirstDirectory(root) ?? projectPath;
                if (root.TryGetProperty("$set", out var set) && set.ValueKind == JsonValueKind.Object)
                {
                    sessionId = String(set, "sessionId") ?? sessionId;
                    projectPath = FirstDirectory(set) ?? projectPath;
                    if (set.TryGetProperty("messages", out var messages) && messages.ValueKind == JsonValueKind.Array)
                        foreach (var message in messages.EnumerateArray()) if (Event(message, sessionId, projectPath) is { } usage) result.Add(usage);
                }
                if (Event(root, sessionId, projectPath) is { } parsed) result.Add(parsed);
            }
            catch { }
        }
        return result.GroupBy(x => x.EventId).Select(x => x.Last()).ToArray();
    }

    private static IReadOnlyList<AIToolUsageEvent> ParseLegacy(JsonElement root, string fallbackSessionId)
    {
        var sessionId = String(root, "sessionId") ?? fallbackSessionId;
        var projectPath = FirstDirectory(root) ?? "unknown";
        if (!root.TryGetProperty("messages", out var messages) || messages.ValueKind != JsonValueKind.Array) return Array.Empty<AIToolUsageEvent>();
        return messages.EnumerateArray().Select(x => Event(x, sessionId, projectPath)).Where(x => x is not null).Cast<AIToolUsageEvent>()
            .GroupBy(x => x.EventId).Select(x => x.Last()).ToArray();
    }

    private static AIToolUsageEvent? Event(JsonElement value, string sessionId, string projectPath)
    {
        if (String(value, "type") != "gemini" || !value.TryGetProperty("tokens", out var tokens) || tokens.ValueKind != JsonValueKind.Object) return null;
        if (!DateTime.TryParse(String(value, "timestamp"), CultureInfo.InvariantCulture, DateTimeStyles.RoundtripKind, out var timestamp)) return null;
        var input = Integer(tokens, "input");
        var output = Integer(tokens, "output");
        var cached = Integer(tokens, "cached");
        var reasoning = Integer(tokens, "thoughts");
        var total = Integer(tokens, "total");
        if (total <= 0) total = input + output + reasoning;
        if (total <= 0) return null;
        var messageId = String(value, "id") ?? timestamp.Ticks.ToString(CultureInfo.InvariantCulture);
        return new AIToolUsageEvent(timestamp, AIToolKind.GeminiCli, "google", String(value, "model") ?? "unknown",
            input, output, cached, reasoning, total, $"gemini:{sessionId}:{messageId}", sessionId, projectPath);
    }

    private static string? FirstDirectory(JsonElement value)
    {
        if (!value.TryGetProperty("directories", out var dirs) || dirs.ValueKind != JsonValueKind.Array) return null;
        foreach (var item in dirs.EnumerateArray())
            if (item.ValueKind == JsonValueKind.String) return item.GetString();
        return null;
    }

    private static string? String(JsonElement value, string name) =>
        value.ValueKind == JsonValueKind.Object && value.TryGetProperty(name, out var item) && item.ValueKind == JsonValueKind.String ? item.GetString() : null;

    private static long Integer(JsonElement value, string name) =>
        value.TryGetProperty(name, out var item) ? Integer(item) : 0;

    private static long Integer(JsonElement value)
    {
        if (value.TryGetInt64(out var integer)) return integer;
        if (value.TryGetDouble(out var number) && double.IsFinite(number)) return (long)number;
        return value.ValueKind == JsonValueKind.String && long.TryParse(value.GetString(), NumberStyles.Integer, CultureInfo.InvariantCulture, out integer)
            ? integer : 0;
    }
}

public static class OpenCodeMessageParser
{
    public static AIToolUsageEvent? ParseMessageData(string json)
    {
        try
        {
            using var doc = JsonDocument.Parse(json);
            var value = doc.RootElement;
            if (String(value, "role") != "assistant" || !value.TryGetProperty("tokens", out var tokens) ||
                !value.TryGetProperty("time", out var time) || !time.TryGetProperty("created", out var createdElement) ||
                !createdElement.TryGetDouble(out var created)) return null;
            var input = Integer(tokens, "input");
            var output = Integer(tokens, "output");
            var reasoning = Integer(tokens, "reasoning");
            var cached = tokens.TryGetProperty("cache", out var cache) ? Integer(cache, "read") : 0;
            var total = Integer(tokens, "total");
            if (total <= 0) total = input + output + reasoning;
            if (total <= 0) return null;
            var messageId = String(value, "id") ?? created.ToString(CultureInfo.InvariantCulture);
            var sessionId = String(value, "sessionID") ?? "unknown";
            var projectPath = value.TryGetProperty("path", out var path) ? String(path, "cwd") ?? "unknown" : "unknown";
            return new AIToolUsageEvent(DateTimeOffset.FromUnixTimeMilliseconds((long)created).UtcDateTime, AIToolKind.OpenCode,
                String(value, "providerID") ?? "unknown", String(value, "modelID") ?? "unknown", input, output, cached, reasoning,
                total, $"opencode:{messageId}", sessionId, projectPath);
        }
        catch { return null; }
    }

    private static string? String(JsonElement value, string name) =>
        value.ValueKind == JsonValueKind.Object && value.TryGetProperty(name, out var item) && item.ValueKind == JsonValueKind.String ? item.GetString() : null;
    private static long Integer(JsonElement value, string name)
    {
        if (value.ValueKind != JsonValueKind.Object || !value.TryGetProperty(name, out var item)) return 0;
        if (item.TryGetInt64(out var integer)) return integer;
        if (item.TryGetDouble(out var number) && double.IsFinite(number)) return (long)number;
        return item.ValueKind == JsonValueKind.String && long.TryParse(item.GetString(), NumberStyles.Integer, CultureInfo.InvariantCulture, out integer)
            ? integer : 0;
    }
}
